### PR TITLE
Sam harper/gs issue fix

### PR DESCRIPTION
**Pull request by rafaellopesdesa**
Hi Sam,

I am adding the producer to fix the photons. Together in the PR goes the recent central PR argiro/fix-gainswitch-80x and some packages that seems non-related but that were updated with when I merged this and your area in a fresh CMSSW_8_0_25 area.

Cheers, Rafael

### CODE REVIEW

**Comment by Sam-Harper** [comment]
Thanks for this Rafael! 

I'm going to see if I can figure out why we pulled a few extra things in and spend a short time to see if we can avoid doing so.

**Comment by Sam-Harper** [comment]
Okay, I've got a test version of this without the extra unrelated changes but all the changes you made.

GSIssueFixTest

https://github.com/Sam-Harper/cmssw/compare/GSIssueFix...Sam-Harper:GSIssueFixTest?expand=1

However I'm going to see if I can do it a bit better avoiding the cherry picks.

**Comment by rafaellopesdesa** [comment]
Thanks. :)

**Comment by Sam-Harper** [comment]
Okay the rebase had exactly the same effect, meh, it'll do. 

Do you want to resubmit the new branch or shall I just merge the new one without the extra stuff. 

If not, I'll make a few comments on this PR as there are some things I want to change

**Comment by rafaellopesdesa** [comment]
Hi,

Please, go ahead, merge the one without the extra stuff (and feel free to
add changes as you see fit).

I already have extra modifications that I am working over this temp branch
of yours (no problem, I can rebase afterwards). But the sooner you have
your modifications in, the better. FYI: I am working in modifying the PAT
sequence to read from this new collection and only saving the electrons
from the old collection when they were modified. I was also told that MET
has to be updated, but I don't know if I should do that or if this should
be done by JetMET people.

File: RecoEgamma/EgammaTools/interface/GainSwitchTools.h [hunk 1]
```diff
@@ -60,7 +60,7 @@ class GainSwitchTools {
 
 template<bool noZS>
 reco::GsfElectron::ShowerShape 
-GainSwitchTools::redoEcalShowerShape(reco::GsfElectron::ShowerShape showerShape,const reco::SuperClusterRef& superClus, const EcalRecHitCollection* recHits,const CaloTopology* topology,const CaloGeometry* geometry)
```
Comment (Sam-Harper) [inline RecoEgamma/EgammaTools/interface/GainSwitchTools.h:63]: Thanks for fixing the HCAL showershapes. However this function was only meant to apply to the ECAL ones. The HCAL ones are more complicated as 

hcalDepth1OverEcal and hcalDepth2OverEcal (standard showershapes)  use the regression ecal energy and will be automatically corrected here for electrons: http://cmslxr.fnal.gov/source/DataFormats/EgammaCandidates/src/GsfElectron.cc?v=CMSSW_8_0_22

So we cant correct here.

**Pull request closed without merging**